File: kata/parens/parens.cpp
#include <iostream>
#include <vector>

using namespace std;

void print_vector(vector<char> vc) {
  cout<<"print";
  for (int i=0;i<vc.size();i++) {
    cout<<vc[i]<<" ";
  }
}

bool check_parens(string line) {
  vector<char> stack;
  for (int i=0;i<line.length();i++) {
    //cout<<stack;
    //print_vector(stack);
    if (line[i] == '(') {
      stack.push_back(line[i]);
    }
    else if (line[i] == ')') {
      if (stack.size() >0) {
        if (stack[stack.size()-1] == '(') {
            stack.pop_back();
        }
      }
      else {
        return false;
      }
    }
  }
  if (stack.size() > 0) {
    return false;
  }
  else {
    return true;
  }
}


int main() {
  for (std::string line; std::getline(std::cin, line);) {
    if (check_parens(line)) {
      cout<<"yes"<<endl;
    }
    else {
      cout<<"no"<<endl;
    }
  }
}

File: kata/parens/parens.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"container/list"
)

func checkParens(line string) bool {
	stack := list.New()
	for _,s :=range line {
		if s == '(' {
			stack.PushBack(s)
		} else {
			if s == ')' {
				if stack.Len() >0 {
					if stack.Back().Value == '(' {
						stack.Remove(stack.Back())
					}
				} else {
					return false
				}
			}
		}
	}

	if stack.Len() >0 {
		return false
	} else {
		return true
	}
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanLines)
	for scanner.Scan() {
		str := scanner.Text()
		fmt.Println(str)
		if checkParens(str) {
			fmt.Println("yes")
		} else {
			fmt.Println("no")
		}
	}
}
